### PR TITLE
Deprecate `auto_select_gpus`

### LINKED ISSUE #10535: [RFC] Deprecate `auto_select_gpus` from Trainer constructor
## Proposed refactoring or deprecation

Deprecate `auto_select_gpus` from Trainer constructor:
https://github.com/PyTorchLightning/pytorch-lightning/blob/7a9a08c5d3ca4699fb439f691c40e1320b37507a/pytorch_lightning/trainer/trainer.py#L209-L212

### Motivation

Users can achieve the same functionality by setting the `accelerator` and `devices` flags - we can simplify by only having one way of doing this.

cc @justusschock @awaelchli @rohitgr7 @tchaton @kaushikb11 @borda

### LINKED ISSUE #13079: RFC: Deprecate  `auto_select_gpus` Trainer argument
## Proposed refactor

Deprecate the `Trainer(auto_select_gpus=True|False)` option and just enable it always when `devices=int` gets selected. 

### Motivation

- The feature is misleading, we don't actually have an algorithm to select the gpus automatically in a smart way. A confusion by user was raised recently: #13012
- The flag only applies to the GPU accelerator. There are no equivalent flags for tpus, ipus, etc.
- What auto_select_gpus does is such a niche use case that it's almost not worth talking about. The implementation just runs through all available GPUs and tests if it can place a tensor in memory. This essentially tests wether the GPU is in exclusive mode or not. [See the implementation here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/tuner/auto_gpu_select.py).
- Internally, the flag is framed as "tuning" as the functions are under the tuner module, but this does not fall under the term "tuning" in my opinion.

### Pitch

Keep the feature, but simply enable it by default and remove the flag from the Trainer. It is still useful to have the check against exclusivity, for example on managed clusters, and I can't think of a reason why it would be undesired. This would only apply when passing `devices=int` and not when indices get passed. 

### Alternatives

Keep it and come up with sophisticated methods to determine whether a GPU should be selected or not based on memory profile, utilization etc. This IMO is not feasible because not enough information is available before training.

### Additional context

#13012

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #13079
Fixes #10535
Alternative to #13356

Before
```py
# Picks the first two available GPUs
Trainer(devices=2, auto_select_gpus=True)
```

After
```py
# Picks the first two available GPUs
from pytorch_lightning.accelerators import find_usable_cuda_gpus

Trainer(devices=find_usable_cuda_gpus)
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @tchaton @justusschock @awaelchli @borda

### CODE REVIEW

File: docs/source-pytorch/common/trainer.rst [hunk 1]
```diff
@@ -317,6 +317,10 @@ before any training.
 auto_select_gpus
 ^^^^^^^^^^^^^^^^
 
+.. warning:: ``auto_select_gpus`` has been deprecated in v1.9.0 and will be removed in v1.10.0.
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:320]: Normally we just remove mentions of deprecated code from the docs

File: docs/source-pytorch/accelerators/gpu_basic.rst [hunk 2]
```diff
@@ -88,10 +88,25 @@ The table below lists examples of possible input formats and how they are interp
...
-    For more details see the :doc:`trainer guide <../common/trainer>`.
+Find usable CUDA devices
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you want to run several experiments at the same time on your machine, for example for a hyperparameter sweep, then you can
+use the following utility function to pick GPU indices that are "accessible", without having to change your code every time.
+
+.. code-block:: python
+
+    from lightning.pytorch.accelerators import find_usable_cuda_gpus
```
Comment (carmocca) [inline docs/source-pytorch/accelerators/gpu_basic.rst:100]: Can this be

```suggestion
    from lightning.pytorch.accelerators.cuda import find_usable_cuda_gpus
```

File: src/lightning_lite/accelerators/__init__.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 from lightning_lite.accelerators.accelerator import Accelerator  # noqa: F401
 from lightning_lite.accelerators.cpu import CPUAccelerator  # noqa: F401
 from lightning_lite.accelerators.cuda import CUDAAccelerator  # noqa: F401
+from lightning_lite.accelerators.cuda import find_usable_cuda_gpus  # noqa: F401
```
Comment (carmocca) [inline src/lightning_lite/accelerators/__init__.py:16]: IMO this utility should be left at the `cuda` level

File: src/lightning_lite/accelerators/cuda.py [hunk 4]
```diff
@@ -78,10 +78,61 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
         )
 
 
-def _get_all_available_cuda_gpus() -> List[int]:
+def find_usable_cuda_gpus(num_gpus: int = -1) -> List[int]:
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:81]: I suggest using "devices" instead of "gpus" for consistency with the trainer flag

```suggestion
def find_usable_cuda_devices(num_devices: int = -1) -> List[int]:
```

If you agree, this should also be changed in the docstrings etc

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -558,8 +558,17 @@ def _set_devices_flag_if_auto_passed(self) -> None:
...
+            rank_zero_deprecation(
+                "The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v1.10.0."
+                " Please use the function `pytorch_lightning.accelerators.find_usable_cuda_gpus` instead."
+            )
         if self._auto_select_gpus and isinstance(self._gpus, int) and isinstance(self.accelerator, CUDAAccelerator):
-            self._devices_flag = pick_multiple_gpus(self._gpus)
+            self._devices_flag = pick_multiple_gpus(
+                self._gpus,
+                # we already show a deprecation message when user sets Trainer(auto_select_gpus=...)
+                __show_deprecatio=False,
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:570]:
```suggestion
                __show_deprecation=False,
```

**Review by carmocca — COMMENTED** [review]
Early review

File: docs/source-pytorch/common/trainer.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source-pytorch/common/trainer.rst:320]: The entire section?

File: src/lightning_lite/accelerators/__init__.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/__init__.py:16]: The import would be too long:
`lightning.fabric.accelerators.cuda`

File: docs/source-pytorch/common/trainer.rst (hunk 1 above)
Comment (carmocca) [reply docs/source-pytorch/common/trainer.rst:320]: Yes. All mentions of the deprecated functionality. This way new users don't even bother with deprecated code.

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: src/lightning_lite/accelerators/cuda.py [hunk 6]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
         )
 
 
-def _get_all_available_cuda_gpus() -> List[int]:
+def find_usable_cuda_devices(num_devices: int = -1) -> List[int]:
+    """Returns a list of all available and usable CUDA GPUs.
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:82]:
```suggestion
    """Returns a list of all available and usable CUDA devices.
```

File: src/lightning_lite/accelerators/cuda.py [hunk 7]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+
+    A GPU is considered usable if we can successfully move a tensor to the device, and this is what this function
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:84]:
```suggestion
    A CUDA device is considered usable if we can successfully move a tensor to the device, and this is what this function
```

File: src/lightning_lite/accelerators/cuda.py [hunk 8]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+    tests for each GPU on the system until the target number of usable GPUs is found.
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:85]:
```suggestion
    tests for each device on the system until the target number of usable devices is found.
```

File: src/lightning_lite/accelerators/cuda.py [hunk 9]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+
+    A subset of GPUs on the system might be used by other processes, and if the GPU is configured to operate in
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:87]:
```suggestion
    A subset of the CUDA devices on the system might be used by other processes, and if the device is configured to operate in
```

File: src/lightning_lite/accelerators/cuda.py [hunk 10]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+    'exclusive' mode (configurable by the admin), then only one process is allowed to occupy it.
+
+    Args:
+        num_devices: The number of GPUs you want to request. By default, this function will return as many as there are
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:91]:
```suggestion
        num_devices: The number of CUDA devices you want to request. By default, this function will return as many as there are
```

File: src/lightning_lite/accelerators/cuda.py [hunk 11]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+            usable GPUs available.
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:92]:
```suggestion
            usable devices available.
```

File: src/lightning_lite/accelerators/cuda.py [hunk 12]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+    Warning:
+        If multiple processes call this function at the same time, there can be race conditions in the case where
+        both processes determine that the device is unoccupied, leading into one of them crashing later on.
     """
-    Returns:
-         A list of all available CUDA GPUs
+    visible_devices = _get_all_visible_cuda_devices()
+    if not visible_devices:
+        raise ValueError(
+            f"You requested to find {num_devices} GPUs but there are no visible CUDA devices on this machine."
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:101]:
```suggestion
            f"You requested to find {num_devices} devices but there are no visible CUDA devices on this machine."
```

File: src/lightning_lite/accelerators/cuda.py [hunk 13]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+        )
+    if num_devices > len(visible_devices):
+        raise ValueError(
+            f"You requested to find {num_devices} GPUs but this machine only has {len(visible_devices)} GPUs."
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:105]:
```suggestion
            f"You requested to find {num_devices} devices but this machine only has {len(visible_devices)} CUDA devices."
```

File: src/lightning_lite/accelerators/cuda.py [hunk 14]
```diff
@@ -78,10 +78,63 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+            continue
+
+        available_devices.append(gpu_idx)
+        if len(available_devices) == num_devices:
+            # exit early if we found the right number of GPUs
+            break
+
+    if len(available_devices) != num_devices:
+        raise RuntimeError(
+            f"You requested to find {num_devices} GPUs but only {len(available_devices)} are currently available."
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:125]:
```suggestion
            f"You requested to find {num_devices} CUDA devices but only {len(available_devices)} devices are currently available."
```

**Review by carmocca — COMMENTED** [review]
A CUDA device is a GPU but a GPU might not be a CUDA device (e.g. MPS)

File: src/lightning_lite/accelerators/cuda.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning_lite/accelerators/cuda.py:82]: I will add some of these, but from the context it should be very clear that GPU means NVIDIA CUDA. 
GPU is not a forbidden word. We can use it and people are very familiar with it. There is nothing wrong with the current description.

> A CUDA device is a GPU but a GPU might not be a CUDA device (e.g. MPS)

You can take this further, into the absurd:

A CUDA device is a GPU but a GPU might not be a CUDA device (e.g. MPS), unless it is an AMD device then it is again pretending to be a CUDA GPU in PyTorch unless you are on MacOS with an AMD GPU then it is actually the MPS device in PyTorch.

**Review by carmocca — APPROVED** [review]
